Unit tests in "crypto/fips1402" passing on RHEL 8.6 with BC FIPS appr…

mposolda opened the pull request:
…oved mode. Cleanup

Closes #13128

The aim of this PR is to have all unit tests passing in the `crypto` module on the FIPS enabled RHEL 8.6 system with FIPS enabled OpenJDK. The "crypto" modules contain some unit tests of the crypto utilities and test that crypto algorithms work as expected. Previously the tests were available inside `keycloak-common` and `keycloak-core` modules. Added the note to `docs/fips.md` about how to run the tests.

- The `JWETest` was failing in the BouncyCastle approved mode due the code from class `KeyEncryptionJWEAlgorithmProvider` did not work with BCFIPS on FIPS based platform. There is a need to use slightly different API provided by BCFIPS. I forked this part from the Trellix branch https://github.com/keycloak/keycloak/compare/14.0.0...sudeepd:saml-fixes, which already had this sorted (in this PR, it is available in the class `FIPSRsaKeyEncryptionJWEAlgorithmProvider`)

- The unit tests in the `crypto/default` module are using non-fips BC and are not supposed to be executed on the FIPS based platform. So some of the tests are skipped on the FIPS platform (most of others were already skipped before this PR). And vice-versa - tests from `crypto/fips1402` are supposed to be executed on the FIPS based platform only.

- Some more cleanup was done (Removing some thombstones from pom.xml, remove static initialization blocks from `PemUtils` and `CertificateUtils` classes etc).

mposolda committed:
Unit tests in "crypto/fips1402" passing on RHEL 8.6 with BC FIPS approved mode. Cleanup
Closes #13128

stianst commented on core/src/main/java/org/keycloak/jose/jwe/JWERegistry.java:
```diff
@@ -61,7 +46,12 @@ class JWERegistry {
     static JWEAlgorithmProvider getAlgProvider(String alg) {
-        return ALG_PROVIDERS.get(alg);
+        // https://tools.ietf.org/html/rfc7518#page-12
+        if (JWEConstants.DIR.equals(alg)) {
```
Can we come up with a better name for DIR? That is a bit confusing name (I was thinking filesystem dir), at least expand it to DIRECT.

mposolda replied:
+1, constants renamed.

The value of the algorithm comes from the JWE specification, so that one has still the same value: https://datatracker.ietf.org/doc/html/rfc7518#section-4.1

stianst commented on crypto/default/src/test/java/org/keycloak/crypto/def/test/DefaultCryptoRSAVerifierTest.java:
```diff
@@ -1,10 +1,19 @@
+    @Before
+    public void before() {
+        // Run this test just if java is not in FIPS mode
+        Assume.assumeFalse("Java is in FIPS mode. Skipping the test.", Environment.isJavaInFipsMode());
```
Would be good to detach "test mode" from "server mode". That can be done in a follow-up PR for sure.

Basically the tests should be executed explicitly in FIPS mode rather than just rely on the "auto-detect" stuff working properly.

mposolda replied:
+1

Right now it is already possible to use the system property `com.redhat.fips` to explicitly enforce if tests should be triggered in the FIPS mode (with BCFIPS) or in the non-fips mode. The auto-detect is used just if this property is not provided. But it will need to be addressed more properly though.

stianst reviewed (commented):
LGTM, some minor comments

mposolda committed:
Rename JWEConstants.DIR

mposolda commented:
@stianst Thanks! Replied and addressed your comment with the DIRECT (added separate commit on top for the easier review, but will squash this before merge).
